src/main/java/com/iota/iri/hash/Kerl.java

```diff
@@ -85,63 +91,108 @@ public void squeeze(final int[] trits, int offset, int length) {
     public static BigInteger bigIntFromTrits(final int[] trits, final int offset, final int size) {
 
         BigInteger value = BigInteger.ZERO;
-
-        for (int i = size; i-- > 0; ) {
-
-            value = value.multiply(BigInteger.valueOf(Converter.RADIX)).add(BigInteger.valueOf(trits[offset + i]));
+        int power = 0;
+        for (int i = offset + size - 1; i >= offset; i--) {
+            switch (trits[i]) {
+                case 1: {
+                    int sequence = 0;
+                    while (i - 1 >= offset && trits[i - 1] == 1) {
+                        i--;
+                        sequence++;
+                    }
+                    value = value.multiply(RADIX_POWERS[power + 1 + sequence]).add(RADIX_POSITIVE_ONE_FUNCTION[sequence]);
+                    power = 0;
+                }
+                break;
+                case -1: {
+                    int sequence = 0;
+                    while (i - 1 >= offset && trits[i - 1] == -1) {
+                        i--;
+                        sequence++;
+                    }
+                    value = value.multiply(RADIX_POWERS[power + 1 + sequence]).add(RADIX_NEGATIVE_ONE_FUNCTION[sequence]);
+                    power = 0;
+                }
+                break;
+                default:
+                    power++;
+            }
+        }
+        // do leftover pow
+        if (power > 0) {
+            value = value.multiply(RADIX_POWERS[power]);
         }
-
         return value;
     }
 
     public static BigInteger bigIntFromBytes(final byte[] bytes, final int offset, final int size) {
-
-        return new BigInteger(Arrays.copyOfRange(bytes, offset, offset + size));
+        return (bytes.length == offset + size) ? new BigInteger(bytes)
+            : new BigInteger(Arrays.copyOfRange(bytes, offset, offset + size));
     }
```

Comment:
Almost 100% of the time, the size is the array given. So this is an easy optimization to not do a copy.